### DISCUSSION TITLE
What's RLS value for backend apps ?

### DISCUSSION THREAD

**Discussion by smallufo** [supabase/supabase#27401] [Questions] [answered]
For the past few months, I have been receiving notifications from Supabase advising me to enable Row Level Security (RLS). 
However, my application is a Java/Kotlin Spring app with just API endpoints and no JWT or browser involvement. Is it still necessary to enable RLS in this case?

If I enable RLS, will my code need to be modified? If so, such changes would be too cumbersome.

**Comment by GaryAustin1** [comment, marked as answer]
If you never ever release your anon key in public, or use any Supabase REST call (including auth or realtime) from a browser then it does not matter if RLS is on or off.  
It would be safer to enable it and use service_role client calls to bypass it, or direct database calls but not required.

**Comment by m-thomson** [comment]
Hi @GaryAustin1, 

I'm an a similar situations as the OP: I have a couple tables that will be backend only. I'm wondering if I should enabled RLS. Specifically, I'm wondering if there are any performance implications to enabling RLS on massive tables (>100M rows) when I don't "need" it?

I understand why RLS is good, but is there a reason why Supabase doesn't have any concept of table-level security for situations like this? What am I missing?

FYI, I'm self hosting (because of the DB size) but would be willing to use paid Supabase with my self-hosted tables added via FDW. I'm not sure what the implications of that would be though.

Finally, is it possible to disable the RLS notifications if we decide to forgo it?

I greatly appreciate any guidance.

**Reply by GaryAustin1** [reply]
Backend or not both can use RLS.  If you are using the REST client from the backend with anon or authenticated users RLS applies.  If you don't have those users than you can use service_role to completely bypass RLS and do all security in your server.  You don't say how you are access SB from the backend.  Enabling RLS with no policies will not impact performance of service role as it bypasses RLS.

Supabase is Postgres.  There are table permissions that will work on postgres roles (like to not allow anon or authenticated roles) if desired.  You can also block the public schema from anon or authenticated having access, but that would apply to your backend if you use them.

You can also secure the database REST API with a function that runs on every request: https://supabase.com/docs/guides/api/securing-your-api?queryGroups=database-method&database-method=sql&queryGroups=pre-request&pre-request=use-additional-api-key#enforce-additional-rules-on-each-request

I've not seen a documented way to turn off notifications, but I am not constantly up on every change they make.

**Reply by m-thomson** [reply]
Thanks Gary, that's really helpful. 

With regards to enabling RLS and then bypassing it with service_role, I guess my question is more of a PG internals question; What I'm wondering is how much disk/memory overhead is added for RLS on such a massive table. And if all rows will have the same access policy anyways, whether it will be worth it.

Right now, my PG instance performs well (64GB RAM, ~45GB Total DB size on disk) since the data is well cached by the OS. If enabling RLS increases the total database size significantly then I worry performance will suffer since everything may not fit in RAM anymore. 

I guess I'm trying to find out if enabling RLS adds a certain amount of overhead _per row_ that I should be aware of, or if it works another, more efficient way.

I guess I just have to try it and see. 

It would be convenient to use SB/Postgrest to access this table if the overhead of RLS is negligible but I'd be willing to code my own standalone REST api if RLS causes a problem (since the table in question, is internal only).

I'm still new to PG/SB as I'm coming from a MySQL background so sorry I've not been clear. I appreciate your time.

**Reply by GaryAustin1** [reply]
I'm not a Postgres expert but RLS should not add anything to the actual data.  It is similar to an extra WHERE clause on the query.  I does add overhead to the query IF you have a role that needs to meet RLS.  On large tables with RLS being enforced for a role that does not bypass it you do need to be careful of your policies and tune them.  There is a section in the SB docs on tuning.

**Comment by m-thomson** [comment]
>It is similar to an extra WHERE clause on the query

Aha!

That's the missing piece I failed to understand. This makes **way** more sense than the model I had constructed in my head whereby every row has some sort of unique permissions metadata. That changes everything about my understanding of RLS and what I'm trying to accomplish. This is perfect. Thanks again for you help!
